### PR TITLE
[#96] Add go_router as a main navigator

### LINKED ISSUE #96: Add go_router as a main navigator
## Why

[Go_router](https://gorouter.dev/) is initially created by Chris Sells - PM at Flutter as a side project. This package now becomes the coolest package ever in [pub.dev](https://pub.dev/packages/go_router), and it's officially maintained by the Flutter team. It makes the developer life easier, and some of the features that meet our needs could be: deep linking and dynamic linking, handling logged-in state, routes and nested routes, etc...

For more info, please check the [go_router page](https://gorouter.dev/getting-started) and watch the video to see how beautiful it is 🤗 

Currently we are manually creating a custom router (in Varun project, it requires a lot of refactoring when implementing deeplink feature), applying `go_router` to the upcoming projects from scratch would be better for us.

If you agree with this, please give a 👍 , otherwise 👎 .

## Who Benefits?

Developers

### PR DESCRIPTION
**Pull request by manh-t**
https://github.com/nimblehq/flutter_templates/issues/96

## What happened 👀

- Add go_router package.
- Setup `GoRouter` for template app.
- Create `Page1Screen` and `Page2Screen` to demonstrate the usage of GoRouter.

Close #96 

## Insight 📝

- [Go Router page](https://gorouter.dev/)
- Tutorial: https://www.raywenderlich.com/28987851-flutter-navigator-2-0-using-go_router
- The different between `go` and `push` in go_router: https://codewithandrea.com/articles/flutter-navigation-gorouter-go-vs-push/

## Proof Of Work 📹

https://user-images.githubusercontent.com/60863885/178691800-2613d0be-0c66-4cee-a79e-f285ea0f9d78.mp4

### CODE REVIEW

**Comment by markgravity** [comment]
@manh-t 
As this https://github.com/nimblehq/compass/pull/980 is official, can you update the `What happened` to follow this new guideline, please? 🙏

File: .template/lib/main.dart [hunk 1]
```diff
@@ -65,10 +84,55 @@ class MyHomePage extends StatelessWidget {
...
+
+  @override
+  Widget build(BuildContext context) {
+    return Scaffold(
+      appBar: AppBar(
+        title: const Text("Page 2"),
+      ),
+      body: Center(
+          child: ElevatedButton(
+        onPressed: () => context.push('/page3'),
```
Comment (markgravity) [inline .template/lib/main.dart:133]: What do you think about adding an example about pushing with parameters? 🤔

**Review by markgravity — COMMENTED** [review]
The rest LGTM 👍

File: .template/lib/main.dart (hunk 1 above)
Comment (manh-t) [reply .template/lib/main.dart:133]: @markgravity I don't want to add too many examples into our template app, demonstrating the `go` and `push` functions are enough IMO. If we want to deep dive into `go_router` features, we can create an app in [flutter-showcases](https://github.com/nimblehq/flutter-showcases) repo for that purpose I think 😄

**Comment by manh-t** [comment]
@markgravity the description has been updated 🙏

**Review by markgravity — APPROVED** [review]
Nice works 👍

File: .template/lib/main.dart [hunk 2]
```diff
@@ -65,10 +84,55 @@ class MyHomePage extends StatelessWidget {
             Text(
               FlutterConfig.get('SECRET'),
               style: const TextStyle(color: Colors.black, fontSize: 24),
-            )
+            ),
+            const SizedBox(height: 24),
+            ElevatedButton(
+              onPressed: () => context.go('/page2'),
```
Comment (luongvo) [inline .template/lib/main.dart:90]: to be reused here...

File: .template/lib/main.dart [hunk 3]
```diff
@@ -65,10 +84,55 @@ class MyHomePage extends StatelessWidget {
...
+        title: const Text("Page 3"),
+      ),
+      body: Center(
+          child: ElevatedButton(
+        onPressed: () => context.pop(),
+        child: const Text("Back"),
+      )),
+    );
+  }
+}
```
Comment (luongvo) [inline .template/lib/main.dart:118]: page3 at the end of this class?

File: .template/lib/main.dart (hunk 1 above)
Comment (luongvo) [inline .template/lib/main.dart:133]: and here 🤔

File: .template/lib/main.dart [hunk 4]
```diff
@@ -2,34 +2,53 @@ import 'package:flutter/material.dart';
...
+        builder: (BuildContext context, GoRouterState state) =>
+            const MyHomePage(),
+        routes: [
+          GoRoute(
+            path: 'page2',
+            builder: (BuildContext context, GoRouterState state) =>
+                const Page2Screen(),
+          ),
+          GoRoute(
+            path: 'page3',
```
Comment (luongvo) [inline .template/lib/main.dart:30]: Should we define some route path constants?

**Review by luongvo — COMMENTED** [review]
rest lgtm

File: .template/lib/main.dart (hunk 2 above)
Comment (manh-t) [reply .template/lib/main.dart]: Fixed in 5a789646d0597aa3c6719c36c9fe667e256ec780

File: .template/lib/main.dart (hunk 3 above)
Comment (manh-t) [reply .template/lib/main.dart]: Fixed in 5a789646d0597aa3c6719c36c9fe667e256ec780

File: .template/lib/main.dart (hunk 1 above)
Comment (manh-t) [reply .template/lib/main.dart]: Fixed in 5a789646d0597aa3c6719c36c9fe667e256ec780

File: .template/lib/main.dart (hunk 4 above)
Comment (manh-t) [reply .template/lib/main.dart]: Fixed in 5a789646d0597aa3c6719c36c9fe667e256ec780

File: .template/lib/main.dart [hunk 5]
```diff
@@ -65,10 +88,55 @@ class MyHomePage extends StatelessWidget {
...
+              child: const Text("Navigate to Page 2"),
+            ),
           ],
         ),
       ),
     );
   }
 }
+
+class Page2Screen extends StatelessWidget {
```
Comment (sleepylee) [inline .template/lib/main.dart:104]:
```suggestion
class SecondScreen extends StatelessWidget {
```

a Page is a reference to a WEB page, while a Screen is more relative to a mobile SCREEN, besides, we should avoid the number in our names. Instead, how about calling by its order, i.e `first` - `second` - `third` in the written format?
Also, do we really need to have a 3rd one? I feel like unnecessary, remember that we should keep this template as simple as possible, right? For demo purposes, you should do it in example/, don't make a mess in this backbone template.

Comment (manh-t) [reply .template/lib/main.dart:104]: @sleepylee agree, I will update 👌

Comment (manh-t) [reply .template/lib/main.dart]: Updated 🙏

**Review by sleepylee — APPROVED** [review]

**Review by luongvo — APPROVED** [review]
💪

**Pull request merged**